lib/bolt/cli.rb

```diff
@@ -379,710 +309,563 @@ def validate(options)
                 "Option '--env-var' can only be specified when running a command or script"
         end
       end
-    end
 
-    def handle_parser_errors
-      yield
-    rescue OptionParser::MissingArgument => e
-      raise Bolt::CLIError, "Option '#{e.args.first}' needs a parameter"
-    rescue OptionParser::InvalidArgument => e
-      raise Bolt::CLIError, "Invalid parameter specified for option '#{e.args.first}': #{e.args[1]}"
-    rescue OptionParser::InvalidOption, OptionParser::AmbiguousOption => e
-      raise Bolt::CLIError, "Unknown argument '#{e.args.first}'"
-    end
+      unless %w[guide module project secret].include?(options[:subcommand]) ||
+             %w[convert new show].include?(options[:action]) ||
+             options[:plan_hierarchy]
+        target_opts   = options.slice(*TARGETING_OPTIONS)
+        target_string = "'--targets', '--rerun', or '--query'"
 
-    def puppetdb_client
-      plugins.puppetdb_client
+        if target_opts.length > 1
+          raise Bolt::CLIError, "Only one targeting option #{target_string} can be specified"
+        elsif target_opts.empty? && options[:subcommand] != 'plan'
+          raise Bolt::CLIError, "Command requires a targeting option: #{target_string}"
+        end
+      end
     end
 
-    def plugins
-      @plugins ||= Bolt::Plugin.setup(config, pal, analytics)
-    end
+    # Execute a Bolt command. The +options+ hash includes the subcommand and
+    # action to be run, as well as any additional arguments and options for the
+    # command.
+    #
+    # @param options [Hash] The CLI options.
+    #
+    def execute(options)
+      with_signal_handling do
+        with_error_handling do
```

Comment:
I think this with_error_handling can be moved into the `exe/bolt` rescue statement so we don't have to wrap parse and execute in it, and that can just call `cli.fatal_error`.